File: android-app/app/src/main/java/com/bookdigest/app/data/remote/ApiService.kt
package com.bookdigest.app.data.remote

import com.bookdigest.app.data.model.*
import retrofit2.Response
import retrofit2.http.*

interface ApiService {
    
    // Authentication
    @POST("auth/register")
    suspend fun register(@Body request: RegisterRequest): Response<AuthResponse>
    
    @POST("auth/login")
    suspend fun login(@Body request: LoginRequest): Response<AuthResponse>
    
    @GET("auth/profile")
    suspend fun getProfile(): Response<UserResponse>
    
    // Books
    @GET("books")
    suspend fun getBooks(
        @Query("page") page: Int = 1,
        @Query("limit") limit: Int = 20,
        @Query("category") category: String? = null,
        @Query("isPremium") isPremium: Boolean? = null
    ): Response<BooksResponse>
    
    @GET("books/{id}")
    suspend fun getBookById(@Path("id") bookId: String): Response<BookDetailResponse>
    
    @GET("books/featured")
    suspend fun getFeaturedBooks(): Response<BooksResponse>
    
    @GET("books/search")
    suspend fun searchBooks(
        @Query("q") query: String,
        @Query("page") page: Int = 1
    ): Response<BooksResponse>
    
    // Favorites
    @POST("books/{id}/favorite")
    suspend fun toggleFavorite(@Path("id") bookId: String): Response<FavoriteResponse>
    
    @GET("books/favorites/me")
    suspend fun getFavorites(): Response<BooksResponse>
    
    // Progress
    @POST("books/{id}/progress")
    suspend fun updateProgress(
        @Path("id") bookId: String,
        @Body progress: ProgressRequest
    ): Response<ProgressResponse>
    
    @GET("books/{id}/progress")
    suspend fun getProgress(@Path("id") bookId: String): Response<ProgressResponse>
    
    // Reviews
    @POST("books/{id}/reviews")
    suspend fun addReview(
        @Path("id") bookId: String,
        @Body review: ReviewRequest
    ): Response<ReviewResponse>
    
    @GET("books/{id}/reviews")
    suspend fun getReviews(
        @Path("id") bookId: String,
        @Query("page") page: Int = 1
    ): Response<ReviewsResponse>
    
    // Categories
    @GET("categories")
    suspend fun getCategories(): Response<CategoriesResponse>
    
    @GET("categories/{slug}/books")
    suspend fun getCategoryBooks(
        @Path("slug") slug: String,
        @Query("page") page: Int = 1
    ): Response<BooksResponse>
    
    // User Stats
    @GET("users/stats")
    suspend fun getUserStats(): Response<UserStatsResponse>
    
    @GET("users/history")
    suspend fun getReadingHistory(@Query("page") page: Int = 1): Response<HistoryResponse>
    
    // Payments
    @POST("payments/create-checkout-session")
    suspend fun createCheckoutSession(@Body request: CheckoutRequest): Response<CheckoutResponse>
    
    @GET("payments/subscription-status")
    suspend fun getSubscriptionStatus(): Response<SubscriptionResponse>
}

File: ios-app/BookDigest/Data/Network/APIService.swift
import Foundation
import Alamofire

class APIService {
    static let shared = APIService()
    
    private let baseURL = "https://api.bookdigest.com/api"
    private var authToken: String? {
        KeychainManager.shared.getToken()
    }
    
    private init() {}
    
    // MARK: - Authentication
    
    func register(email: String, password: String, firstName: String?, lastName: String?) async throws -> AuthResponse {
        let parameters: [String: Any] = [
            "email": email,
            "password": password,
            "firstName": firstName ?? "",
            "lastName": lastName ?? ""
        ]
        
        return try await request(
            endpoint: "/auth/register",
            method: .post,
            parameters: parameters
        )
    }
    
    func login(email: String, password: String) async throws -> AuthResponse {
        let parameters: [String: Any] = [
            "email": email,
            "password": password
        ]
        
        return try await request(
            endpoint: "/auth/login",
            method: .post,
            parameters: parameters
        )
    }
    
    func getProfile() async throws -> UserResponse {
        return try await request(endpoint: "/auth/profile", method: .get)
    }
    
    // MARK: - Books
    
    func getBooks(page: Int = 1, category: String? = nil, isPremium: Bool? = nil) async throws -> BooksResponse {
        var parameters: [String: Any] = ["page": page, "limit": 20]
        if let category = category { parameters["category"] = category }
        if let isPremium = isPremium { parameters["isPremium"] = isPremium }
        
        return try await request(
            endpoint: "/books",
            method: .get,
            parameters: parameters
        )
    }
    
    func getBookById(_ id: String) async throws -> BookDetailResponse {
        return try await request(endpoint: "/books/\(id)", method: .get)
    }
    
    func getFeaturedBooks() async throws -> BooksResponse {
        return try await request(endpoint: "/books/featured", method: .get)
    }
    
    func searchBooks(query: String, page: Int = 1) async throws -> BooksResponse {
        let parameters: [String: Any] = ["q": query, "page": page]
        return try await request(
            endpoint: "/books/search",
            method: .get,
            parameters: parameters
        )
    }
    
    func toggleFavorite(bookId: String) async throws -> FavoriteResponse {
        return try await request(endpoint: "/books/\(bookId)/favorite", method: .post)
    }
    
    func getFavorites() async throws -> BooksResponse {
        return try await request(endpoint: "/books/favorites/me", method: .get)
    }
    
    // MARK: - Progress
    
    func updateProgress(bookId: String, progress: ProgressRequest) async throws -> ProgressResponse {
        return try await request(
            endpoint: "/books/\(bookId)/progress",
            method: .post,
            parameters: progress.toDictionary()
        )
    }
    
    func getProgress(bookId: String) async throws -> ProgressResponse {
        return try await request(endpoint: "/books/\(bookId)/progress", method: .get)
    }
    
    // MARK: - Reviews
    
    func addReview(bookId: String, rating: Int, comment: String?) async throws -> ReviewResponse {
        let parameters: [String: Any] = [
            "rating": rating,
            "comment": comment ?? ""
        ]
        
        return try await request(
            endpoint: "/books/\(bookId)/reviews",
            method: .post,
            parameters: parameters
        )
    }
    
    // MARK: - Categories
    
    func getCategories() async throws -> CategoriesResponse {
        return try await request(endpoint: "/categories", method: .get)
    }
    
    // MARK: - User Stats
    
    func getUserStats() async throws -> UserStatsResponse {
        return try await request(endpoint: "/users/stats", method: .get)
    }
    
    // MARK: - Payments
    
    func createCheckoutSession(planType: String) async throws -> CheckoutResponse {
        let parameters: [String: Any] = ["planType": planType]
        return try await request(
            endpoint: "/payments/create-checkout-session",
            method: .post,
            parameters: parameters
        )
    }
    
    func getSubscriptionStatus() async throws -> SubscriptionResponse {
        return try await request(endpoint: "/payments/subscription-status", method: .get)
    }
    
    // MARK: - Generic Request Method
    
    private func request<T: Decodable>(
        endpoint: String,
        method: HTTPMethod,
        parameters: [String: Any]? = nil
    ) async throws -> T {
        let url = baseURL + endpoint
        
        var headers: HTTPHeaders = [
            "Content-Type": "application/json",
            "Accept": "application/json"
        ]
        
        if let token = authToken {
            headers["Authorization"] = "Bearer \(token)"
        }
        
        return try await withCheckedThrowingContinuation { continuation in
            AF.request(
                url,
                method: method,
                parameters: parameters,
                encoding: method == .get ? URLEncoding.default : JSONEncoding.default,
                headers: headers
            )
            .validate()
            .responseDecodable(of: T.self) { response in
                switch response.result {
                case .success(let value):
                    continuation.resume(returning: value)
                case .failure(let error):
                    continuation.resume(throwing: error)
                }
            }
        }
    }
}
